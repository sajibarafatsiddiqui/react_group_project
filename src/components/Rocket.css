.container {
  display: flex;
}

.rocket-img {
  max-width: 20vw;
  margin: 20px;
}

.reserve-btn {
  background-color: blue;
  color: white;
  border: none;
  border-radius: 4px;
  height: 30px;
}

.cancel-btn {
  background-color: whitesmoke;
  color: grey;
  border: 1px solid grey;
  border-radius: 4px;
  height: 30px;
}

.badge {
  background-color: rgb(19, 181, 187);
  color: white;
  border-radius: 4px;
  height: 20px;
}
