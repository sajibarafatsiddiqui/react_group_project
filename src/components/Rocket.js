import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { reserveRocket, cancelReservation } from 'redux/rockets/rocketsSlice';
import './Rocket.css';

const Rocket = ({ rockets }) => {
  const dispatch = useDispatch();

  const handleReserve = (id) => {
    dispatch(reserveRocket(id));
    localStorage.setItem(`rocket-${id}`, 'reserved');
  };

  const handleCancelReservation = (id) => {
    dispatch(cancelReservation(id));
    localStorage.removeItem(`rocket-${id}`);
  };

  const isReserved = (id) => localStorage.getItem(`rocket-${id}`) === 'reserved';

  return (
    <div data-testid="rockets-container">
      {rockets.map((rocket) => (
        <div key={rocket.id} className="container">
          {rocket.flickr_images && rocket.flickr_images.length > 0 && (
            <img
              src={rocket.flickr_images[0]}
              alt={rocket.name}
              className="rocket-img"
            />
          )}
          <div className="text-wrapper">
            <div>
              <h2>{rocket.name}</h2>
              <p>
                {isReserved(rocket.id) && <span className="badge">Reserved</span>}
                {' '}
                {rocket.description}
              </p>
            </div>
            {isReserved(rocket.id) ? (
              <button
                type="button"
                className="cancel-btn"
                onClick={() => handleCancelReservation(rocket.id)}
              >
                Cancel reservation
              </button>
            ) : (
              <button
                type="button"
                className="reserve-btn"
                onClick={() => handleReserve(rocket.id)}
              >
                Reserve Rocket
              </button>
            )}
          </div>
        </div>
      ))}
    </div>
  );
};

export default Rocket;

Rocket.propTypes = {
  rockets: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired,
      flickr_images: PropTypes.arrayOf(PropTypes.string),
    }),
  ).isRequired,
};
