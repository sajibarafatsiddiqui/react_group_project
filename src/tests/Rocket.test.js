import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import Rocket from '../components/Rocket';
import store from '../redux/store';

describe('Tests for <Rocket /> component', () => {
  const rockets = [{
    id: '1',
    name: 'Rocket 1',
    description: 'Rocket Desc 1',
    flickr_images: ['1', '2', '3'],
  }, {
    id: '2',
    name: 'Rocket 2',
    description: 'Rocket Desc 2',
    flickr_images: ['1', '2', '3'],
  }];

  const wrapper = (component) => (
    <Provider store={store}>
      <>
        {component}
      </>
    </Provider>
  );

  it('Should render two rockets', () => {
    render(wrapper(<Rocket rockets={rockets} />));
    expect(screen.queryByTestId('rockets-container').children.length).toBe(2);
  });

  it('Should match the snapshot', () => {
    expect(renderer.create(wrapper(
      <Rocket rockets={rockets} />,
    )).toJSON()).toMatchSnapshot();
  });
});
