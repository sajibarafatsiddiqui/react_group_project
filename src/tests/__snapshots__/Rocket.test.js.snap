// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <Rocket /> component Should match the snapshot 1`] = `
<div
  data-testid="rockets-container"
>
  <div
    className="container"
  >
    <img
      alt="Rocket 1"
      className="rocket-img"
      src="1"
    />
    <div
      className="text-wrapper"
    >
      <div>
        <h2>
          Rocket 1
        </h2>
        <p>
           
          Rocket Desc 1
        </p>
      </div>
      <button
        className="reserve-btn"
        onClick={[Function]}
        type="button"
      >
        Reserve Rocket
      </button>
    </div>
  </div>
  <div
    className="container"
  >
    <img
      alt="Rocket 2"
      className="rocket-img"
      src="1"
    />
    <div
      className="text-wrapper"
    >
      <div>
        <h2>
          Rocket 2
        </h2>
        <p>
           
          Rocket Desc 2
        </p>
      </div>
      <button
        className="reserve-btn"
        onClick={[Function]}
        type="button"
      >
        Reserve Rocket
      </button>
    </div>
  </div>
</div>
`;
